Mathlib/Order/PrimeSeparator.lean

```diff
@@ -0,0 +1,155 @@
+
+  -- Suppose that a₁ ∉ J, a₂ ∉ J. We need to prove that a₁ ⊔ a₂ ∉ J.
+  rw [isPrime_iff_mem_or_mem]
+  intros a₁ a₂
+  contrapose!
+  intro ⟨ha₁, ha₂⟩
+
+  -- Consider the ideals J₁, J₂ generated by J ∪ {a₁} and J ∪ {a₂}, respectively.
+  let J₁ := J ⊔ principal a₁
+  let J₂ := J ⊔ principal a₂
+
+  -- For each i, Jᵢ is an ideal that contains J, I and aᵢ, and is not equal to J.
+  have JsubJ₁ : J ≤ J₁ := le_sup_left
+  have JsubJ₂ : J ≤ J₂ := le_sup_left
+  have IJ₁ : I ≤ J₁ := le_trans IJ' le_sup_left
+  have IJ₂ : I ≤ J₂ := le_trans IJ' le_sup_left
+  have a₁J₁ : a₁ ∈ J₁ := mem_of_subset_of_mem (le_sup_right : _ ≤ J ⊔ _) mem_principal_self
+  have a₂J₂ : a₂ ∈ J₂ := mem_of_subset_of_mem (le_sup_right : _ ≤ J ⊔ _) mem_principal_self
+  have J₁J : ↑J₁ ≠ Jset := by refine ne_of_mem_of_not_mem' a₁J₁ ha₁
+  have J₂J : ↑J₂ ≠ Jset := by refine ne_of_mem_of_not_mem' a₂J₂ ha₂
+
+  -- Therefore, since J is maximal, we must have Jᵢ ∉ S.
+  have J₁S : ↑J₁ ∉ S := fun h => J₁J (Jmax J₁ h JsubJ₁)
+  have J₂S : ↑J₂ ∉ S := fun h => J₂J (Jmax J₂ h JsubJ₂)
+
+  -- Since Jᵢ is an ideal that contains I, we have that Jᵢ is not disjoint from F.
+  have J₁F : ¬ (Disjoint (F : Set α) J₁) := by
+    intro hdis
+    apply J₁S
+    simp only [le_eq_subset, mem_setOf_eq, SetLike.coe_subset_coe, S]
+    exact ⟨J₁.isIdeal, IJ₁, hdis⟩
+
+  have J₂F : ¬ (Disjoint (F : Set α) J₂) := by
+    intro hdis
+    apply J₂S
+    simp only [le_eq_subset, mem_setOf_eq, SetLike.coe_subset_coe, S]
+    exact ⟨J₂.isIdeal, IJ₂, hdis⟩
+
+  -- Thus, pick cᵢ ∈ F ∩ Jᵢ.
+  obtain ⟨c₁, ⟨c₁F, c₁J₁⟩⟩ := Set.not_disjoint_iff.1 J₁F
+  obtain ⟨c₂, ⟨c₂F, c₂J₂⟩⟩ := Set.not_disjoint_iff.1 J₂F
+
+  -- Using the definition of Jᵢ, we can pick bᵢ ∈ J such that cᵢ ≤ bᵢ ⊔ aᵢ.
+  obtain ⟨b₁, ⟨b₁J, cba₁⟩⟩ := (mem_ideal_sup_principal a₁ c₁ J).1 c₁J₁
+  obtain ⟨b₂, ⟨b₂J, cba₂⟩⟩ := (mem_ideal_sup_principal a₂ c₂ J).1 c₂J₂
+
+  -- Since J is an ideal, we have b := b₁ ⊔ b₂ ∈ J.
+  let b := b₁ ⊔ b₂
+  have bJ : b ∈ J := sup_mem b₁J b₂J
+
+  have ineq : c₁ ⊓ c₂ ≤ b ⊔ (a₁ ⊓ a₂) :=
+  -- Now we compute: b ⊔ (a₁ ⊓ a₂) = (b ⊔ a₁) ⊓ (b ⊔ a₂) by distributivity
+  -- and b ⊔ aᵢ ≥ bᵢ ⊔ aᵢ by definition of b and monotonicity of ⊔.
+  -- Thus, b ⊔ (a₁ ⊓ a₂) ≥ (b₁ ⊔ a₁) ⊓ (b₂ ⊔ a₂) ≥ c₁ ⊓ c₂,
+  -- with the last inequality holding by monotonicity of inf.
+  calc
+    c₁ ⊓ c₂ ≤ (b₁ ⊔ a₁) ⊓ (b₂ ⊔ a₂) := inf_le_inf cba₁ cba₂
+    _       ≤ (b  ⊔ a₁) ⊓ (b  ⊔ a₂) := by
+      apply inf_le_inf <;> apply sup_le_sup_right; exact le_sup_left; exact le_sup_right
+    _       = b ⊔ (a₁ ⊓ a₂) := (sup_inf_left b a₁ a₂).symm
+
+  -- Note that c₁ ⊓ c₂ ∈ F, since c₁ and c₂ are both in F and F is a filter.
+  -- Since F is an upper set, it now follows that b ⊔ (a₁ ⊓ a₂) ∈ F.
+  have ba₁a₂F : b ⊔ (a₁ ⊓ a₂) ∈ F := PFilter.mem_of_le ineq (PFilter.inf_mem c₁F c₂F)
+
+  -- Now, if we would have a₁ ⊓ a₂ ∈ J, then, since J is an ideal and b ∈ J, we would also get
+  -- b ⊔ (a₁ ⊓ a₂) ∈ J. But this contradicts that J is disjoint from F.
+
+  intro ha₁a₂
```

Comment:
You could probably do `apply not_not_intro JF`

Reply:
This did not work for me, the goal at this point is a₁ ⊓ a₂ ∉ J which is not of the required form ¬ ¬ p.

Reply:
This was just a suggestion to avoid proving an intermediate lemma (and have a more backwards-type proof). No problem besides that.